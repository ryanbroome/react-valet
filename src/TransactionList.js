import React, { useEffect, useState, useContext } from "react";
import { useHistory, Link } from "react-router-dom";
import { FormGroup, Form, Input, Label, Button } from "reactstrap";

import ValetApi from "./api/Api";
// import AddVehicleForm from "./AddVehicleForm";

import TransactionSearchForm from "./TransactionSearchForm";
import TransactionCard from "./TransactionCard";
import ThemeContext from "./ThemeContext";

// import VehicleCard from "./VehicleCard";

const TransactionList = () => {
  const { userDetail, token } = useContext(ThemeContext);
  // const [mobileSearchTerm, setMobileSearchTerm] = useState("''");
  const [transactions, setTransactions] = useState([]);
  const history = useHistory();

  useEffect(
    function fetchTransactionsWhenMounted() {
      async function fetchTransactions() {
        try {
          const res = await ValetApi.getAllTransactionsByLocationAndStatus(userDetail.locationId, "parked");
          setTransactions(res.data.transactions);
        } catch (err) {
          console.log(err);
        }
      }
      fetchTransactions();
    },
    [userDetail.locationId]
  );

  const handleSearchByMobile = async (locationId, mobile) => {
    try {
      const searchRes = await ValetApi.getActiveTransactionsByMobile(locationId, mobile);
      setTransactions([...searchRes.data.transactions]);
    } catch (err) {
      console.log(err);
    }
  };

  const handleReset = async (locationId, status) => {
    try {
      const resetRes = await ValetApi.getAllTransactionsByLocationAndStatus(locationId, status);
      setTransactions([...resetRes.data.transactions]);
      history.push("/transactions/active");
    } catch (err) {
      console.error(`Error resetting getAllTransactionsByLocationAndStatus`);
    }
  };

  const checkout = async (username, vehicleId, transactionId) => {
    try {
      await ValetApi.checkoutVehicle(vehicleId);
      await ValetApi.parkOne(username);

      setTransactions(transactions.filter((trans) => trans.transactionId !== transactionId));
      history.push("/addVehicle");
    } catch (err) {
      console.error(`Error parkOne for username : ${username} vehicleId : ${vehicleId}`);
    }
  };

  //TODO doesnt work need to pass in vehicleID, transactionID not sure I want to delete a vehicle
  const handleRemove = async (vehicleId, transactionId) => {
    await ValetApi.deleteVehicle(vehicleId);
    await ValetApi.deleteTransaction(transactionId);
    // remove from local storage
    history.push("/transactions/active");
  };

  // *What to render if token present or not
  function loggedIn(transactions) {
    return (
      <div>
        <TransactionSearchForm
          search={handleSearchByMobile}
          reset={handleReset}
        />
        {transactions.length > 0
          ? transactions.map((trans) => (
              <div key={`Transaction-${trans.transactionId}`}>
                <TransactionCard
                  transaction={trans}
                  checkout={checkout}
                />
              </div>
            ))
          : "false"}
      </div>
    );
  }

  function loggedOut() {
    return (
      <div>
        <p>Please Login to see your ActiveGarage</p>
      </div>
    );
  }

  return token ? loggedIn(transactions) : loggedOut();
};

export default TransactionList;
