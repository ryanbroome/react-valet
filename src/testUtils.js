import React from "react";
import UserContext from "./auth/UserContext";
// import { render } from "@testing-library/react";
// import { MemoryRouter } from "react-router";
import App from "../src/App";
// import TransactionList from "./TransactionList";

const MockUserContext = ({ children }) => {
  const mockToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InRlc3R1c2VyIiwibG9jYXRpb25JZCI6MSwiaXNBZG1pbiI6ZmFsc2UsImlhdCI6MTcxMzA3NTM4Mn0.dXlnawfzk6wlsgpouoSp58QLvlOOSJFC03iq_FXab-I";
  const mockUserDetail = {
    userId: 1,
    vehicleId: 1,
    locationId: 1,
    transactionId: 1,
    ticketNum: "1",
    mobile: "5553334444",
    color: "red",
    make: "honda",
    username: "testuser",
  };
  const mockLocation = {};
  const mockVehicle = {};
  const mockTransaction = {};

  return <UserContext.Provider value={{ token: mockToken, userDetail: mockUserDetail, location: mockLocation, vehicle: mockVehicle, transaction: mockTransaction, update: App.update }}>{children}</UserContext.Provider>;
};

export { MockUserContext };
