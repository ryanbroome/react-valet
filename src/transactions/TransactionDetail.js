import { Card, CardBody, CardTitle, CardSubtitle, ListGroup, ListGroupItem, CardText, Button } from "reactstrap";

import React, { useEffect, useState, useContext } from "react";
import { useParams, useHistory, Link } from "react-router-dom";
import ValetApi from "../api/Api";

const TransactionDetail = () => {
  const { id } = useParams();
  const [transaction, setTransaction] = useState();

  useEffect(
    function fetchTransactionsAtMount() {
      async function fetchTransactions() {
        try {
          const transactionRes = await ValetApi.getTransactionById(id);
          setTransaction(transactionRes.data.transactions);
        } catch (err) {
          console.log(err);
        }
      }
      fetchTransactions();
    },
    [id]
  );

  function viewTransaction(transaction) {
    return (
      <Card
        color="dark"
        body
        className="text-center">
        <CardBody>
          <CardSubtitle>{transaction.ticketNum}</CardSubtitle>
          <ListGroup flush>
            <ListGroupItem>Site: {transaction.sitename}</ListGroupItem>
            <ListGroupItem>Site ID: {transaction.locationId}</ListGroupItem>
            <ListGroupItem>Trans ID: {transaction.transactionId}</ListGroupItem>
            <ListGroupItem>Ticket:{transaction.ticketNum}</ListGroupItem>
            <ListGroupItem>Trans Time: {transaction.transactionTime}</ListGroupItem>
            <ListGroupItem>Mobile: {transaction.mobile}</ListGroupItem>
            <ListGroupItem>Vehicle ID: {transaction.vehicleId}</ListGroupItem>
            <ListGroupItem>Status: {transaction.vehicleStatus}</ListGroupItem>
            <ListGroupItem>In Time: {transaction.checkIn}</ListGroupItem>
            <ListGroupItem>Out Time: {transaction.checkOut}</ListGroupItem>
            <ListGroupItem>Color: {transaction.color}</ListGroupItem>
            <ListGroupItem>Make: {transaction.make}</ListGroupItem>
            <ListGroupItem>Damages: {transaction.damages}</ListGroupItem>
            <ListGroupItem>Notes: {transaction.notes}</ListGroupItem>
            <ListGroupItem>Valet ID: {transaction.userId}</ListGroupItem>
            <ListGroupItem>
              Valet: {transaction.firstName} {transaction.lastName}
            </ListGroupItem>
            <ListGroupItem>Last: {transaction.lastName}</ListGroupItem>
            <ListGroupItem>Phone: {transaction.phone}</ListGroupItem>
            <ListGroupItem>Email: {transaction.email}</ListGroupItem>
            <ListGroupItem>TotalParked: {Number(transaction.totalParked) > 0 ? transaction.totalParked : "None"}</ListGroupItem>
          </ListGroup>
        </CardBody>
      </Card>
    );
  }

  function showSpinner() {
    return (
      <div
        className="spinner-border text-primary"
        role="status">
        <span className="visually-hidden"></span>
      </div>
    );
  }
  return transaction ? viewTransaction(transaction) : showSpinner();
};

export default TransactionDetail;
