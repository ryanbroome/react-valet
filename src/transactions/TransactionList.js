import React, { useEffect, useState, useContext } from "react";
import { useHistory } from "react-router-dom";

import ValetApi from "../api/Api";

import TransactionSearchForm from "../forms/TransactionSearchForm";
import TransactionCard from "./TransactionCard";
import UserContext from "../auth/UserContext";

/** Show Transactions for active user location
 *
 * **/
const TransactionList = () => {
  const history = useHistory();
  const { userDetail, token, error, setError } = useContext(UserContext);
  const [transactions, setTransactions] = useState([]);

  useEffect(function fetchTransactionsWhenMounted() {
    async function fetchTransactions() {
      try {
        const res = await ValetApi.getAllTransactionsByLocationAndStatus(userDetail.locationId, "parked");
        setTransactions(res.data.transactions);
      } catch (err) {
        setError(err);
        console.error(err);
      }
    }
    fetchTransactions();
  }, []);

  const handleSearchByMobile = async (locationId, mobile) => {
    try {
      const searchRes = await ValetApi.getActiveTransactionsByMobile(locationId, mobile);
      setTransactions([...searchRes.data.transactions]);
    } catch (err) {
      setError(err);
      console.log(err);
    }
  };

  const handleReset = async (locationId, status) => {
    try {
      const resetRes = await ValetApi.getAllTransactionsByLocationAndStatus(locationId, status);
      setTransactions([...resetRes.data.transactions]);
      history.push("/transactions/active");
    } catch (err) {
      setError(err);
      console.error(err);
    }
  };

  const checkout = async (username, vehicleId, transactionId) => {
    try {
      await ValetApi.checkoutVehicle(vehicleId);
      await ValetApi.parkOne(username);

      setTransactions(transactions.filter((trans) => trans.transactionId !== transactionId));
      history.push("/addVehicle");
    } catch (err) {
      setError(err);
      console.error(err);
    }
  };

  const handleLostKeys = async (locationId, userId) => {
    try {
      const searchRes = await ValetApi.lostKeys(locationId, userId);
      setTransactions([...searchRes.data.transactions]);
    } catch (err) {
      setError(err);
      console.log(err);
    }
  };

  function loggedIn(transactions) {
    return (
      <div>
        {error && <div>{error}</div>}
        <TransactionSearchForm
          search={handleSearchByMobile}
          reset={handleReset}
          lostKeys={handleLostKeys}
        />

        <div className="container-fluid row wrap">
          {transactions.length > 0 ? (
            transactions.map((trans) => (
              <div
                className="col-sm"
                key={`Transaction-Container-${trans.transactionId}`}>
                <TransactionCard
                  key={`TransactionCard-${trans.transactionId}`}
                  transaction={trans}
                  checkout={checkout}
                />
              </div>
            ))
          ) : (
            <div
              className="spinner-border text-primary"
              role="status">
              <span className="visually-hidden"></span>
            </div>
          )}
        </div>
      </div>
    );
  }

  function loggedOut() {
    return (
      <div>
        {error && <div style={{ color: "red" }}>{error}</div>}
        <p>Please Login to see your ActiveGarage</p>
      </div>
    );
  }

  return token ? loggedIn(transactions) : loggedOut();
};

export default TransactionList;
