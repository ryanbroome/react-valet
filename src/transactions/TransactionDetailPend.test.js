import React from "react";
import { render } from "@testing-library/react";
import TransactionDetail from "./TransactionDetail";
import { MemoryRouter, Route } from "react-router-dom";
import TransactionList from "./TransactionList";

import { MockUserContext } from "../testUtils";

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <MockUserContext children={<TransactionList />}></MockUserContext>
    </MemoryRouter>
  );
});

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <UserProvider>
        <TransactionDetail />
      </UserProvider>
    </MemoryRouter>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter initialEntries={["/TransactionDetail/ibm"]}>
      <UserProvider>
        <Route path="/TransactionDetail/:handle">
          <TransactionDetail />
        </Route>
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot with logo", function () {
  const transaction = {
    userId: 1,
    vehicleId: 1,
    locationId: 1,
    transactionId: 1,
    ticketNum: "1",
    mobile: "5553334444",
    color: "red",
    make: "honda",
    username: "testuser",
  };
  const { asFragment } = render(
    <MemoryRouter>
      <TransactionDetail transaction={transaction} />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
