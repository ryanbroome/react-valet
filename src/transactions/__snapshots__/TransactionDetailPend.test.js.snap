// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot with logo 1`] = `
<DocumentFragment>
  <div
    class="spinner-border text-primary"
    role="status"
  >
    <span
      class="visually-hidden"
    />
  </div>
</DocumentFragment>
`;
